tests/catch/unit/memory/MemUtils.hh

```diff
@@ -0,0 +1,638 @@
+ * return a pair of pointers due to the device registered allocation case, we need to keep track of
+ * the pointer to host memory to be able to unregister and free it
+ */
+template <typename T>
+static inline std::pair<T*, T*> initMemory(allocType type, memType memType, MultiDData& data) {
+  size_t dataH = data.height == 0 ? 1 : data.height;
+  size_t dataD = data.depth == 0 ? 1 : data.depth;
+  std::pair<T*, T*> retPtr{};
+  // check different types of allocation
+  switch (type) {
+    case allocType::deviceMalloc:
+      retPtr = deviceMallocHelper<T>(memType, data.width, dataH, dataD, data.pitch);
+      break;
+
+    case allocType::hostMalloc:
+      retPtr = hostMallocHelper<T>(data.width, dataH, dataD, data.pitch);
+      break;
+
+    case allocType::hostRegisted:
+      retPtr = hostRegisteredHelper<T>(data.width, dataH, dataD, data.pitch);
+      break;
+
+    case allocType::devRegistered:
+      retPtr = devRegisteredHelper<T>(data.width, dataH, dataD, data.pitch);
+      break;
+
+    default:
+      REQUIRE(false);
+      break;
+  }
+  return retPtr;
+}
+
+// set of helper functions to tidy the nested switch statements
+template <typename T>
+static inline void deviceMallocCopy(memType memType, T* aPtr, T* hostMem, size_t dataW,
+                                    size_t dataH, size_t dataD, size_t& dataPitch) {
+  size_t elementSize = sizeof(T);
+  size_t sizeInBytes = elementSize * dataW * dataH * dataD;
+  switch (memType) {
+    case memType::hipMem3D: {
+      hipMemcpy3DParms params{};
+      params.kind = hipMemcpyDeviceToHost;
+      params.srcPos = make_hipPos(0, 0, 0);
+      params.srcPtr = make_hipPitchedPtr(aPtr, dataPitch, dataW, dataH);
+      params.dstPos = make_hipPos(0, 0, 0);
+      params.dstPtr = make_hipPitchedPtr(hostMem, dataW, dataW, dataH);
+
+      hipExtent extent;
+      extent.width = dataW;
+      extent.height = dataH;
+      extent.depth = dataD;
+
+      params.extent = extent;
+
+      HIP_CHECK(hipMemcpy3D(&params));
+      break;
+    }
+
+    case memType::hipMem2D:
+      HIP_CHECK(hipMemcpy2D(hostMem, dataW * elementSize, aPtr, dataPitch, dataW, dataH,
+                            hipMemcpyDeviceToHost));
+      break;
+
+    default:
+      HIP_CHECK(hipMemcpy(hostMem, aPtr, sizeInBytes, hipMemcpyDeviceToHost));
+      break;
+  }
+}
+
+template <typename T>
+static inline void hostCopy(memType memType, T* aPtr, T* hostMem, size_t dataW, size_t dataH,
+                            size_t dataD, size_t& dataPitch) {
+  size_t elementSize = sizeof(T);
+  size_t sizeInBytes = elementSize * dataW * dataH * dataD;
+  switch (memType) {
+    case memType::hipMem3D: {
+      hipMemcpy3DParms params{};
+      params.kind = hipMemcpyHostToHost;
+      params.srcPos = make_hipPos(0, 0, 0);
+      params.dstPos = make_hipPos(0, 0, 0);
+      params.srcPtr = make_hipPitchedPtr(aPtr, dataPitch, dataW, dataH);
+      params.dstPtr = make_hipPitchedPtr(hostMem, dataW, dataW, dataH);
```

Comment:
To avoid copy-paste everywhere the concept of functions and lambdas have been invented.

Reply:
Creating ```hipMemcpy3DParms``` has been moved into separate function.